Implement 'env' package to handle environment variables in cosign

xmudrii opened the pull request:
#### Summary

This PR implements a new `env` package to handle environment variables in cosign. This PR/implementation targets only `COSIGN_` environment variables, but it can be extended to handle other environment variables as well. This implementation is a combination of easy and hard mode described here: https://github.com/sigstore/cosign/issues/2236#issuecomment-1241050406

The `env` package implements:

- `Variable` type which is used to define an environment variable
- `VariableOpts` type which is used to describe an environment variable
- A slice with known environment variables and their descriptions (used for printing)
- `Getenv(Variable) string` function which is a wrapper around `os.Getenv` for given `Variable`
- `LookupEnv(Variable) (string, bool)` function which is a wrapper around `os.LookupEnv` for given `Variable`
- `PrintEnv()` function which is used to print all known variables, their values, and descriptions

All `os.Getenv`/`os.LookupEnv` calls on `COSIGN_` variables are replaced with `env.Getenv` and `env.LookupEnv`. There are some other `os.Getenv` calls on non-cosign variables, but those will be handled in a follow up PR (if this turns out to be an approach we want to follow).

This PR also modifies `cosign env` command with some additional features. First of all, the command is now using `env.PrintEnv()` to print the known cosign variables instead of searching Environ. However, we still search Environ for `SIGSTORE_` and `TUF_` variables. Additionally, **there are two new flags**:

- `--show-descriptions` (default `true`) -- used to enable descriptions for known cosign variables
- `--show-sensitive-values` (default `false`) -- by default, `cosign env` is going to show asterisk signs if variable is marked as sensitive. Turning this flag on will show values for sensitive variables

Example 1 (with descriptions and hidden sensitive values, default):

```shell
COSIGN_PASSWORD=test ./cosign env
```

```
# COSIGN_DOCKER_MEDIA_TYPES to be used with registries that do not support OCI media types
COSIGN_DOCKER_MEDIA_TYPES=""
# COSIGN_EXPERIMENTAL enables experimental cosign features
COSIGN_EXPERIMENTAL=""
# COSIGN_PASSWORD overrides password inputs with this value
COSIGN_PASSWORD="******"
# COSIGN_PKCS11_MODULE_PATH is PKCS11 module-path
COSIGN_PKCS11_MODULE_PATH=""
# COSIGN_PKCS11_PIN to be used if PKCS11 PIN is not provided
COSIGN_PKCS11_PIN=""
# COSIGN_REPOSITORY can be used to store signatures in an alternate location
COSIGN_REPOSITORY=""
```

Example 2 (with sensitive values):

```shell
COSIGN_PASSWORD=test ./cosign env --show-sensitive-values
```

```
# COSIGN_DOCKER_MEDIA_TYPES to be used with registries that do not support OCI media types
COSIGN_DOCKER_MEDIA_TYPES=""
# COSIGN_EXPERIMENTAL enables experimental cosign features
COSIGN_EXPERIMENTAL=""
# COSIGN_PASSWORD overrides password inputs with this value
COSIGN_PASSWORD="test"
# COSIGN_PKCS11_MODULE_PATH is PKCS11 module-path
COSIGN_PKCS11_MODULE_PATH=""
# COSIGN_PKCS11_PIN to be used if PKCS11 PIN is not provided
COSIGN_PKCS11_PIN=""
# COSIGN_REPOSITORY can be used to store signatures in an alternate location
COSIGN_REPOSITORY=""
```

One missing piece that is mentioned in the issue, but is not implemented, is a linter to forbid using `os.Getenv` and related functions. This is something that we need to discuss and investigate a bit more. Also, there are other environment variables such as `SIGSTORE_` and `TUF_` that are not handled, but this is something that can be discussed and handled later. I believe this PR should be a solid foundation for any further work on this topic.

Fixes part of https://github.com/sigstore/cosign/issues/2236

#### Release Note

- Implement `env` package used to register and handle environment variables used by cosign
- `cosign env` now prints all registered `COSIGN_` environment variables
- Add `--show-descriptions` (default `true`) flag to `cosign env` used to show description for each registered `COSIGN_` environment variable
- Add `--show-sensitive-values` (default `false`) flag to `cosign env` used to hidh values for `COSIGN_` environment variables marked as sensitive

#### Documentation

I'm not sure, but I guess mostly like yes.

xmudrii commented:
cc @znewman01 PTAL ^^ :slightly_smiling_face:

xmudrii commented:
I have an issue with running `make docgen`:

```
go run -tags pivkey,pkcs11key,cgo ./cmd/help/
# pkg-config --cflags  -- libpcsclite
Package libpcsclite was not found in the pkg-config search path.
Perhaps you should add the directory containing `libpcsclite.pc'
to the PKG_CONFIG_PATH environment variable
No package 'libpcsclite' found
pkg-config: exit status 1
make: *** [Makefile:176: docgen] Error 2
```

Any idea how can I fix this? I'm running Ubuntu 22.04.

cpanato commented:
> I have an issue with running `make docgen`:
> 
> ```
> go run -tags pivkey,pkcs11key,cgo ./cmd/help/
> # pkg-config --cflags  -- libpcsclite
> Package libpcsclite was not found in the pkg-config search path.
> Perhaps you should add the directory containing `libpcsclite.pc'
> to the PKG_CONFIG_PATH environment variable
> No package 'libpcsclite' found
> pkg-config: exit status 1
> make: *** [Makefile:176: docgen] Error 2
> ```
> 
> Any idea how can I fix this? I'm running Ubuntu 22.04.

you need to install the `libpcsclite-dev` - `apt install libpcsclite-dev`

znewman01 commented on pkg/cosign/env/env.go:
```diff
@@ -0,0 +1,120 @@
+	}
+)
+
+func Getenv(name Variable) string {
```
Should this enforce that `name` is in `environmentVariables`? IMO okay to `panic` if not. (Same for `LookupEnv`)

Also should it enforce that `name` starts with `COSIGN_`?

xmudrii replied:
> Should this enforce that name is in environmentVariables? IMO okay to panic if not. (Same for LookupEnv)

I wasn't sure if it's okay to panic or not, but I think in this case it's acceptable. I'll add this change.

> Also should it enforce that name starts with COSIGN_?

I thought that we might want to use this mechanism for some non-cosign environment variables. Considering that we're not doing this right now, I'm going to enforce this and we can always remove it if needed.

znewman01 replied:
> I'm going to enforce this and we can always remove it if needed.

+1 to that thought

znewman01 commented on pkg/cosign/env/env.go:
```diff
@@ -0,0 +1,120 @@
+	return os.LookupEnv(name.String())
+}
+
+func PrintEnv(showDescription, showSensitive bool) {
```
Perhaps this should be moved to `cli/`? WDYT?

xmudrii replied:
The reason why I didn't want this in `cli/` is to avoid publicly exposing `environmentVariables` slice. This is because I want to avoid someone changing the slice outside the env package. If this is important, I can workaround it with some getter or I can make the slice public.

znewman01 replied:
Good thought. I think it's prudent to avoid letting callers mutate that slice.

I think it's worth it to add a getter, though.

xmudrii replied:
This is now moved to `cli/`.

znewman01 commented on pkg/cosign/env/env.go:
```diff
@@ -0,0 +1,120 @@
+
+type Variable string
+
+type VariableOpts struct {
```
nit: types might be nice here so users know to pass `=1` or `=true` vs. `=my password`. You could do this directly in the descriptions too.

xmudrii replied:
Great idea, will add it!

znewman01 reviewed (commented):
Great idea to redact sensitive information by default 👍

xmudrii commented:
@znewman01 I should have addressed all your comments. I'm trying to configure `forbidigo` and after that the PR should be ready.

xmudrii commented:
@znewman01 I added `forbidigo` to golangci-lint. There are some places where we still use `os.Getenv` with non-cosign environment variables. I handled that in two ways:

- if the file/package has many os.Getenv/LookupEnv calls, ignore that file or package to `.golangci.yml` file
- if the file/package has only one or few os.Getenv/LookupEnv calls, use in-line `//nolint:forbidigo` comment

Please take a look at this and let me know if this approach is okay or if I should change anything.

znewman01 reviewed (commented):
I'm confused about how verification will work. Verifiers need to reconstruct the entry. Do they have enough information to do so, and can we do so unambiguously?

Specifically:

- Every `timestamp_verification_data` in a given bundle should contain signatures for the *same* Rekor entry, right?
- If so, why is there a `KindData` in *each* `timestamp_verification_data`?
- There are cases where the artifact + bundle doesn't have the information we need to reconstruct the Rekor entry (e.g., if we only have RFC 3161 timestamps, there's no KindVersion), right?

And two broader questions. First, how entry-type-aware should applications/client libraries be? The Alpine Rekor entries include metadata from the PKGINFO. AFAICT, these are just unvalidated metadata and duplicate information from the artifact. To verify the Rekor signature, you'd need to parse the .apk file and format the entry yourself, which feels like a lot of work. Then, the signature lives inside the artifact.

I feel like I'm missing some context on the design philosophy of Rekor entries. Here's how I think about it:

- A Rekor entry contains:
  - some signed data (or a reference to such data)
  - the signature over that data (or the signature could be in theh artifact itself)
  - optionally with unverified metadata that can be used for searching.
- The entry can contain the data itself or just a hash; it can contain the signature itself, or the signature can live inside the data.
- Users can use the unverified metadata for searching, but it should be considered untrusted.
- After verification, users can trust the artifact along with *some* of the other metadata (exactly which metadata depends on the verification procedure used from the type).
- They still cannot trust the unsigned metadata, unless they reconstruct it from the artifact (at which point, they should just use the metadata from the artifact).

Why aren't we putting the whole entry into `bundle.content`? IIRC, two reasons. First, to avoid sticking unvalidated public keys in the bundle. Second, it's a lot of duplication with the VerificationMaterial. Are we better off fixing the Rekor entry types themselves to remove the VerificationMaterial? Rekor wouldn't be able to validate the entries, but it already can't, in a lot of cases.

--------------------------------------------------------------------------------

Sorry that I keep coming back with more issues, but fresh eyes always help me see a little better. Despite my comments I don't *actually* want to block this forever :smile:. But I also think it's critical to get this right. Could we push *something* marked experimental pretty soon (with a self-destruct clock, to prevent it from being entrenched), then try to build client libraries with it? I think that will help us shake out the rought edges.

Also, you may have noticed that I'm always going to err on the side of wanting more types. Feel free to push back on that!

znewman01 commented on .golangci.yml:
```diff
@@ -41,6 +51,16 @@ issues:
+  - path: pkg/cosign/env
+    linters:
+    - forbidigo
+  - path: pkg/providers/
```
I actually think most of these are worth documenting. IMO the goal of `cosign env` is to ask "what from my environment is possibly affecting Cosign" or to ask "what knobs do I have to affect Cosign's behavior?"

(Same for `pkg/cosign/git/gitlab`.)

WDYT?

(I think it's reasonable to address that in a later PR, before closing the accompanying issue.)

xmudrii replied:
> I actually think most of these are worth documenting. IMO the goal of cosign env is to ask "what from my environment is possibly affecting Cosign" or to ask "what knobs do I have to affect Cosign's behavior?"

@znewman01 Just to make sure I have a clear understanding -- do you mean that I should also add those variables to the slice of known/registered variables (`environmentVariables` in the `env` package)? As a result of that, `cosign env` is also going to print variables like `GITHUB_TOKEN`, `GITLAB_TOKEN`, etc...

In that case, we have to remove checking if the variable starts with `COSIGN_` (https://github.com/sigstore/cosign/pull/2322#discussion_r991726750), but I think it makes sense to do that.

znewman01 replied:
Precisely, I think you should add them all to `environmentVariables`.

(And yes, I think the `COSIGN_` prefix check doesn't make sense anymore. Sorry I asked you to add it and now am asking you to take it away!)

xmudrii replied:
Fixed.

znewman01 commented on cmd/cosign/cli/fulcio/fulcioverifier/ctl/verify.go:
```diff
@@ -74,7 +74,7 @@ func ContainsSCT(cert []byte) (bool, error) {
 	// fetch SCT verification key
 	pubKeys := make(map[[sha256.Size]byte]logIDMetadata)
-	rootEnv := os.Getenv(altCTLogPublicKeyLocation)
+	rootEnv := os.Getenv(altCTLogPublicKeyLocation) //nolint:forbidigo
```
IMO this is worth documenting.

znewman01 commented on pkg/cosign/env/env_test.go:
```diff
@@ -0,0 +1,214 @@
+)
+
+func TestGetenv(t *testing.T) {
+	os.Setenv(VariableExperimental.String(), "1")
```
Nit: TBH for a test, I think it might be more readable to hardcode `COSIGN_EXPERIMENTAL`. It makes it clearer what you're testing.

I'll go with your judgment here, though. WYDT?

znewman01 commented on pkg/cosign/tlog.go:
```diff
@@ -110,7 +110,7 @@ func intotoEntry(ctx context.Context, signature, pubKey []byte) (models.Proposed
 func GetRekorPubs(ctx context.Context, rekorClient *client.Rekor) (map[string]RekorPubKey, error) {
 	publicKeys := make(map[string]RekorPubKey)
-	altRekorPub := os.Getenv(altRekorPublicKey)
+	altRekorPub := os.Getenv(altRekorPublicKey) //nolint:forbidigo
```
IMO this is worth including.

znewman01 commented on cmd/cosign/cli/env.go:
```diff
@@ -21,29 +21,41 @@ import (
 }
 
-func getEnv() []string {
+func getExternalEnv() []string {
```
I think it's worth folding this into `PrintEnv`, there are some subtle scenarios to test.

xmudrii replied:
This is done, I'm just not sure if I tested everything expected.

znewman01 replied:
I should have written down what I meant, ha. You covered everything I can think of, *except* if we want to make external environment variables "sensitive" by default. Then we need a test case with `--show-sensitive` and one without

znewman01 commented on cmd/cosign/cli/env.go:
```diff
@@ -21,29 +21,41 @@ import (
+			// Print COSIGN_ environment variables
+			env.PrintEnv(o.ShowDescriptions, o.ShowSensitiveValues)
+
+			// Print external environment variables (SIGSTORE_ and TUF_)
```
Maybe worth printing a comment about these (`# These are environment variables that Cosign doesn't specifically know about, but with prefixes that look like they might be used by some of Cosign's dependencies` or something)

znewman01 replied:
(if `o.ShowDescriptions` is set)

xmudrii replied:
That's a good point, I added something similar.

znewman01 commented on cmd/cosign/cli/env.go:
```diff
@@ -21,29 +21,41 @@ import (
+
+			// Print external environment variables (SIGSTORE_ and TUF_)
+			for _, e := range getExternalEnv() {
 				fmt.Println(e)
```
This is just printing the name, right? Want to avoid sensitive values I think.

xmudrii replied:
No, this is printing `key=value`. We can't really tell if it's sensitive or not in this case.

znewman01 commented on cmd/cosign/cli/env.go:
```diff
@@ -21,29 +21,41 @@ import (
 		// Prefixes to look for. err on the side of showing too much rather
 		// than too little. We'll only output things that have values set.
 		for _, prefix := range []string{
-			"COSIGN_",
```
I wonder if you want to keep `COSIGN_`-prefixed variables, and do something like `if strings.HasPrefix(e, prefix) && e not in env.allEnvVars`? (<--pseudocode)

xmudrii replied:
This is a great idea, I added something like that.

znewman01 commented on internal/pkg/cosign/fulcio/fulcioroots/fulcioroots.go:
```diff
@@ -62,7 +62,7 @@ func initRoots() (*x509.CertPool, *x509.CertPool, error) {
 	var intermediatePool *x509.CertPool
 
-	rootEnv := os.Getenv(altRoot)
+	rootEnv := os.Getenv(altRoot) //nolint:forbidigo
```
IMO worth documenting

znewman01 commented on pkg/cosign/env/env.go:
```diff
@@ -0,0 +1,141 @@
+)
+
+var (
+	environmentVariables = map[Variable]VariableOpts{
```
One thing I'm noticing: this design centralizes the environment variables.

You could also consider a registry, where various parts of Cosign would register their own environment variables. (Closer to where they're used). I think *many* of these are used *a lot of places*, so that approach may not be great :shrug: 

I think this is fine as-is, I just want to make a note so that this decision is explicit.

znewman01 commented on pkg/cosign/env/env.go:
```diff
@@ -0,0 +1,141 @@
+	if _, ok := environmentVariables[name]; !ok {
+		panic(fmt.Sprintf("environment variable %q is not registered in pkg/cosign/env", name.String()))
+	}
+	if !strings.HasPrefix(name.String(), "COSIGN_") {
```
Welp, now that I've asked you to Enforce the `COSIGN_` prefix we might want to start doing `SIGSTORE_` and `TUF_` variables. Can you allowlist all three prefixes? You could also get rid of this altogether in favor of the "is in `environmentVariables`" check

I think in the medium-to-long-term we probably want to do something similar to this pattern in the upstream sigstore/TUF repositories, and then we could really go back to only having `COSIGN_` variables (we'd call `sigstore/pkg/env.LookupEnv` where needed).

xmudrii replied:
I went with some in-between approach here. I added a new field to `VariableOpts` -- `External`. If variable is not external (`External: false`), then it must start with `COSIGN_`, otherwise we don't care.

znewman01 replied:
nice!

znewman01 commented on pkg/cosign/env/env.go:
```diff
@@ -0,0 +1,141 @@
+		panic(fmt.Sprintf("environment variable %q is not registered in pkg/cosign/env", name.String()))
+	}
+	if !strings.HasPrefix(name.String(), "COSIGN_") {
+		panic(fmt.Sprintf("environment varialbe %q must start with COSIGN_ prefix", name.String()))
```
Typo: variable

znewman01 commented on pkg/cosign/env/env_test.go:
```diff
@@ -0,0 +1,214 @@
+	"testing"
+)
+
+func TestGetenv(t *testing.T) {
```
Can we add some tests that `recover()` from the panics of invalid variables?

xmudrii replied:
Added.

znewman01 commented on pkg/cosign/tlog.go:
```diff
@@ -150,7 +150,7 @@ func GetRekorPubs(ctx context.Context, rekorClient *client.Rekor) (map[string]Re
 	// If we have a Rekor client and we've been told to fetch the Public Key from Rekor,
 	// additionally fetch it here.
-	addRekorPublic := os.Getenv(addRekorPublicKeyFromRekor)
+	addRekorPublic := os.Getenv(addRekorPublicKeyFromRekor) //nolint:forbidigo
```
Same here.

znewman01 commented on pkg/cosign/git/github/github.go:
```diff
@@ -44,7 +44,7 @@ func New() *Gh {
 func (g *Gh) PutSecret(ctx context.Context, ref string, pf cosign.PassFunc) error {
 	var httpClient *http.Client
-	if token, ok := os.LookupEnv("GITHUB_TOKEN"); ok {
+	if token, ok := os.LookupEnv("GITHUB_TOKEN"); ok { //nolint:forbidigo
```
IMO worth documenting

znewman01 commented on pkg/blob/load.go:
```diff
@@ -53,7 +53,7 @@ func LoadFileOrURL(fileRef string) ([]byte, error) {
 		case "env://":
 			envVar := parts[1]
-			value, found := os.LookupEnv(envVar)
+			value, found := os.LookupEnv(envVar) //nolint:forbidigo
```
This is a legitimate exception, because it's user-provided, nice find.

Can we add a justification for the lint skip?

```
// Most of Cosign should use `env.LookupEnv` (see #2236) to restrict us to known environment variables (usually `$COSIGN_*`.
// However, in this case, `envVar` is user-provided and not one of the allow-listed env vars.
```

xmudrii replied:
Added.

znewman01 reviewed (changes requested):
This is a really great first contribution to Cosign, thank you so much!

Some changes requested, mostly minor.

xmudrii commented:
@znewman01 Thank you so much for the review! I agree with all those comments. It’s already late for me (I’m in the CEST time zone), so I’ll address those comments beginning of the next week if that’s okay. 🙂

znewman01 commented:
> @znewman01 Thank you so much for the review! I agree with all those comments. It’s already late for me (I’m in the CEST time zone), so I’ll address those comments beginning of the next week if that’s okay. 🙂

No rush! Have a great weekend

xmudrii commented:
@znewman01 I think that I should have addressed all comments. Please let me know if I forgot something or if you have any additional comments. I tried to document all variables as best as I could, but I'm not sure have I got all descriptions and expectations correctly, so please take a look at that.

znewman01 commented on cmd/cosign/cli/env_test.go:
```diff
@@ -16,22 +16,204 @@
+		t.Run(tt.name, func(t *testing.T) {
+			// Set needed environment variables
+			for k, v := range tt.environmentVariables {
+				os.Setenv(k, v)
```
Will this test run into trouble if I have `COSIGN_FOO=1` set when I run the tests?

xmudrii replied:
Yes, it will. This is a great point and something that I missed.

I tried to solve this, I'm not super proud of the fix, but it's in this commit: https://github.com/sigstore/cosign/pull/2322/commits/5c8b0b0af40927cf4b95d0ad767a78158311892a

It's not the best way to solve it, and I'm open to reverting that commit if you don't like the solution.

znewman01 replied:
:shipit:

znewman01 commented on cmd/cosign/cli/env_test.go:
```diff
@@ -16,22 +16,204 @@
+COSIGN_TEST2=""
+# Environment variables below are not registered with cosign,
+# but might still influence cosign's behavior.
+COSIGN_TEST3=abcd
```
Hm...I think the safe thing to do is to treat everything as sensitive unless told otherwise (`--show-sensitive`).

I could just `export COSIGN_MY_ROOT_PASSWORD=hunter2` and it'd get printed out. That's *kind of* my fault, but 🤷. You could imagine a situation where `sigstore/sigstore` adds a new variable for a password and we don't put it into our map by the next Cosign release, which would be actually bad.

It's a little bit user-unfriendly but definitely safe.

xmudrii replied:
I agree it's definitely safer. It should be fixed now. I also added a few test cases for this.

znewman01 commented on cmd/cosign/cli/env_test.go:
```diff
@@ -16,22 +16,204 @@
+COSIGN_TEST2=""
+# Environment variables below are not registered with cosign,
+# but might still influence cosign's behavior.
+COSIGN_TEST3=abcd
```
Do we want quotes here? `COSIGN_TEST3="abcd"`

znewman01 replied:
Or get rid of quotes for the other variables. `env` doesn't show quotes, even if the variable has a space in it. 

```shell
$ env
env | head -n 3
TERM_SESSION_ID=w1t0p0:674E0F26-D173-4E6D-A95C-F26D8DEE667D
SSH_AUTH_SOCK=/private/tmp/com.apple.launchd.g6K1fscsWe/Listeners
LC_TERMINAL_VERSION=3.4.15
ITERM_PROFILE=Hotkey Window
```

Don't feel strongly either way but consistency is nice 🙂

xmudrii replied:
I also think consistency is nice. I removed quotes, it should be the same as for `env` and `os.Environ()` in Go.

znewman01 reviewed (changes requested):
So close! This is looking great. Just a couple small things

xmudrii commented:
@znewman01 I addressed your comments, PTAL when you have some time. :slightly_smiling_face:

znewman01 commented:
Small lint issue, once you fix that I think we can merge!

xmudrii commented:
@znewman01 Fixed. :slightly_smiling_face:

znewman01 commented:
Nice, if tests pass this should merge automatically!

Thank you so much! 🙏  Great first PR, I wish everybody sent one like this 😄

xmudrii commented:
@znewman01 I'm so happy to see this PR merged! Thanks a lot for going through this with me, I really appreciate it!
